src/viur/core/bones/file.py

```diff
@@ -4,97 +4,106 @@
-                for fileName, size, mimetype, customData in callRes:
-                    resDict[fullSrcKey]["files"][fileName] = {
+
+            if call_res := caller(skel, skel["derived"]["files"], params):
+                assert isinstance(call_res, list), "Old (non-list) return value from deriveFunc"
+                res_status[full_src_key] = params_hash
+                for file_name, size, mimetype, custom_data in call_res:
+                    res_files[file_name] = {
                         "size": size,
                         "mimetype": mimetype,
-                        "customData": customData
+                        "customData": custom_data  # TODO: Rename in VIUR4
                     }
 
-    def updateTxn(key, resDict):
-        obj = db.get(key)
-        if not obj:  # File-object got deleted during building of our derives
-            return
-        obj["derived"] = obj.get("derived") or {}
-        obj["derived"]["deriveStatus"] = obj["derived"].get("deriveStatus") or {}
-        obj["derived"]["files"] = obj["derived"].get("files") or {}
-        for k, v in resDict.items():
-            obj["derived"]["deriveStatus"][k] = v["version"]
-            for fileName, fileDict in v["files"].items():
-                obj["derived"]["files"][fileName] = fileDict
-        db.put(obj)
-
-    if resDict:  # Write updated results back and queue updateRelationsTask
-        db.run_in_transaction(updateTxn, key, resDict)
+    def merge_derives(patch_skel):
+        if not patch_skel["derived"]:
+            patch_skel["derived"] = {}
+        patch_skel["derived"] = {"deriveStatus": {}, "files": {}} | patch_skel["derived"]
+        patch_skel["derived"]["deriveStatus"] = patch_skel["derived"]["deriveStatus"] | res_status
+        patch_skel["derived"]["files"] = patch_skel["derived"]["files"] | res_files
+
+    if res_status:  # Write updated results back and queue updateRelationsTask
+
+        skel.patch(values=merge_derives, update_relations=False)
+
         # Queue that updateRelations call at least 30 seconds into the future, so that other ensureDerived calls from
         # the same FileBone have the chance to finish, otherwise that updateRelations Task will call postSavedHandler
         # on that FileBone again - re-queueing any ensureDerivedCalls that have not finished yet.
-        updateRelations(key, time() + 1, "derived", _countdown=30)
-        if refreshKey:
-            def refreshTxn():
-                skel = skeletonByKind(refreshKey.kind)()
-                if not skel.read(refreshKey):
+
+        if refresh_key:
+            def __txn_refresh():
+                _skel = skeletonByKind(refresh_key.kind)()
+                if not _skel.read(refresh_key):
                     return
-                skel.refresh()
-                skel.write(update_relations=False)
+                _skel.refresh()
+                _skel.write(update_relations=False)
+
+            db.run_in_transaction(__txn_refresh)
```

Comment:
Please try this:
```suggestion
            skel = skeletonByKind(refresh_key.kind)()
            skel.patch(lambda skel: skel.refresh(), update_relations=False)
```

Reply:
Then the skel is written twice every time